From 0f8253b45665d02559760b0ed3f3725f2a1c63d1 Mon Sep 17 00:00:00 2001
From: Michael Davidsaver <mdavidsaver@gmail.com>
Date: Sun, 26 Jul 2015 09:28:07 -0400
Subject: tftpDriver: don't free directory node's path string

---
 cpukit/libnetworking/lib/tftpDriver.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/cpukit/libnetworking/lib/tftpDriver.c b/cpukit/libnetworking/lib/tftpDriver.c
index 0d5e536..a774d43 100644
--- a/cpukit/libnetworking/lib/tftpDriver.c
+++ b/cpukit/libnetworking/lib/tftpDriver.c
@@ -512,8 +512,9 @@ static int rtems_tftp_eval_path(
         size_t pathlen = pathnamelen;
         size_t len = currentlen + pathlen;
 
-        current = realloc (current, len + 1);
+        current = malloc (len + 1);
         if (current != NULL) {
+            memcpy (current, currentloc->node_access, currentlen);
             memcpy (current + currentlen, path, pathlen);
             current [len] = '\0';
             if (!rtems_tftp_is_directory (current, len)) {
-- 
2.1.4

